### PR TITLE
kind: Support local registry at "--deploy"

### PR DESCRIPTION
**Pull request by qinqon**
**- What this PR does and why is it needed**
This allow to use kind.sh and `--deploy` with kind local registry.

**- Description for the changelog**
kind: Support local registry at "--deploy"

### CODE REVIEW

File: contrib/kind.sh [hunk 1]
```diff
@@ -603,7 +605,41 @@ create_kind_cluster() {
   if kind get clusters | grep ovn; then
     delete
   fi
+  if [[ "${KIND_LOCAL_REGISTRY}" == true ]]; then
+      # create registry container unless it already exists
+      if [ "$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)" != 'true' ]; then
+        docker run \
```
Comment (maiqueb) [inline contrib/kind.sh:611]: ditto

File: contrib/kind.sh [hunk 2]
```diff
@@ -603,7 +605,41 @@ create_kind_cluster() {
...
+          -d --restart=always -p "127.0.0.1:${reg_port}:5000" --name "${reg_name}" \
+          registry:2
+      fi
+  fi
   kind create cluster --name "${KIND_CLUSTER_NAME}" --kubeconfig "${KUBECONFIG}" --image "${KIND_IMAGE}":"${K8S_VERSION}" --config=${KIND_CONFIG_LCL} --retain
+  # connect the registry to the cluster network if not already connected
+  if [ "$(docker inspect -f='{{json .NetworkSettings.Networks.kind}}' "${reg_name}")" = 'null' ]; then
```
Comment (maiqueb) [inline contrib/kind.sh:618]: ditto

File: contrib/kind.sh [hunk 3]
```diff
@@ -603,7 +605,41 @@ create_kind_cluster() {
...
+    docker network connect "kind" "${reg_name}"
```
Comment (maiqueb) [inline contrib/kind.sh:619]: ditto

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: contrib/kind.sh [hunk 4]
```diff
@@ -603,7 +605,41 @@ create_kind_cluster() {
   if kind get clusters | grep ovn; then
     delete
   fi
+  if [[ "${KIND_LOCAL_REGISTRY}" == true ]]; then
+      # create registry container unless it already exists
+      if [ "$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)" != 'true' ]; then
```
[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh:610]: Right, let me check how it works with podman

[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh:610]: Looks like kind podman support is not GA, https://github.com/kubernetes-sigs/kind/blob/412a2adaeb3d86eadba93010cf72d97c1a0857c6/pkg/cluster/provider.go#L110

Do you know if much of the ovn-k community is using podman to start the kind with it ?

[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh:610]: Can we do a follow up with docker ? the local registry does not work out of the box replacing docker with podman also there is still parts of kind.sh that only works with docker.

Apart from that podman support at kind is not GA https://github.com/kubernetes-sigs/kind/blob/412a2adaeb3d86eadba93010cf72d97c1a0857c6/pkg/cluster/provider.go#L110

[the remark this answers is not in the document]
Comment (maiqueb) [reply contrib/kind.sh:610]: I'm not a maintainer, but if it were up to me, no. This change is useless for people using podman, which is one of the [allowed/advertised](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/kind.md#run-the-kind-deployment-with-podman) container runtimes of the ovn-k kind integration.

Sure, podman support at kind is not GA, but I don't see why that matters since kind is not meant to run production workloads. 

It is possible to load the images using podman. The [multus repo does it](https://github.com/k8snetworkplumbingwg/multus-cni/commit/392726842fc19411834b7bd67fbbe1daf32c1aed).

Finally, afaict, slirp4netns is not needed, since afaiu, ovn-k in kind requires rootfull podman; from [its docs](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/kind.md#run-the-kind-deployment-with-podman):
> To deploy KIND however, you need to start it as root and then copy root's kube config to use it as non-root:

[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh]: Adapter the PR to work with "roorful" podman, thanks!

File: contrib/kind.sh [hunk 5]
```diff
@@ -603,7 +605,37 @@ create_kind_cluster() {
   if kind get clusters | grep ovn; then
     delete
   fi
+  if [[ "${KIND_LOCAL_REGISTRY}" == true ]]; then
+      # create registry container unless it already exists
+      if [ "$($OCI_BIN inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)" != 'true' ]; then
+        $OCI_BIN run \
+          -d --restart=always -p "127.0.0.1:${reg_port}:5000" --name "${reg_name}" \
+          registry:2
+      fi
```
Comment (maiqueb) [inline contrib/kind.sh:614]: maybe encapsulate this in a function ? not a must.

File: contrib/kind.sh [hunk 6]
```diff
@@ -603,7 +605,37 @@ create_kind_cluster() {
...
+  fi
   kind create cluster --name "${KIND_CLUSTER_NAME}" --kubeconfig "${KUBECONFIG}" --image "${KIND_IMAGE}":"${K8S_VERSION}" --config=${KIND_CONFIG_LCL} --retain
+  if [[ "${KIND_LOCAL_REGISTRY}" == true ]]; then
+    
+    # connect the registry to the cluster network if not already connected
+    if [ "$($OCI_BIN inspect -f='{{json .NetworkSettings.Networks.kind}}' "${reg_name}")" = 'null' ]; then
+      $OCI_BIN network connect "kind" "${reg_name}"
+    fi
+
+    # Document the local registry
```
Comment (maiqueb) [inline contrib/kind.sh:624]: document ? I don't follow.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Looks good to me, just a couple of nits.

Would you mind also adding info to `docs/kind.md` explaining how to use this ?

**Review by maiqueb — APPROVED** [review]
Approve on the conceptual level but I would like to see the kind.md docs improved by explaining how to use this from the dev perspective i.e. push to the local registry, force the rollout of the new daemonset pods, etc.

File: docs/kind.md [hunk 7]
```diff
@@ -436,6 +438,36 @@ $ cd ../../contrib/
...
+The kind.sh has the "--deploy" flag to refresh ovn-kubernets without restarting
+kind, so to reload the code just doing the following is enough:
+
+```bash
+# Default options will use kind mechanism to push images directly to the 
+# cluster
+./kind.sh --deploy
+
+# Using the kind local registry will use the conneted local registry to 
+# push images, it make easier to deploy non ovn-k artifacts
```
Comment (maiqueb) [inline docs/kind.md:467]:
```suggestion
# Using the kind local registry will use the connected local registry to
# push images, making it easier to deploy non ovn-k artifacts.
```

File: docs/kind.md [hunk 8]
```diff
@@ -436,6 +438,36 @@ $ cd ../../contrib/
...
+The kind.sh has the "--deploy" flag to refresh ovn-kubernets without restarting
+kind, so to reload the code just doing the following is enough:
+
+```bash
+# Default options will use kind mechanism to push images directly to the 
+# cluster
+./kind.sh --deploy
+
+# Using the kind local registry will use the connected local registry to
+# push images, making easier to deploy non ovn-k artifacts.
```
Comment (maiqueb) [inline docs/kind.md:467]:
```suggestion
# push images, making it easier to deploy non ovn-k artifacts.
```

**Review by maiqueb — APPROVED** [review]

File: docs/kind.md [hunk 9]
```diff
@@ -436,6 +438,36 @@ $ cd ../../contrib/
...
+### Loading ovn-kuberntes changes without restarting kind
+
+At some scenarios user may have other stuff apart from ovn-kubernetes at the
+kind cluster so restarting it involve redploying those.
+
+The kind.sh has the "--deploy" flag to refresh ovn-kubernets without restarting
+kind, so to reload the code just doing the following is enough:
+
+```bash
+# Default options will use kind mechanism to push images directly to the
```
Comment (jcaamano) [inline docs/kind.md:462]:
```
By default, the `kind load` mechanism will be used to update the ovn-kubernetes images
```

File: docs/kind.md [hunk 10]
```diff
@@ -436,6 +438,36 @@ $ cd ../../contrib/
...
+The kind.sh has the "--deploy" flag to refresh ovn-kubernets without restarting
+kind, so to reload the code just doing the following is enough:
+
+```bash
+# Default options will use kind mechanism to push images directly to the 
+# cluster
+./kind.sh --deploy
+
+# Using the kind local registry will use the connected local registry to
+# push images, making it easier to deploy non ovn-k artifacts.
```
Comment (jcaamano) [inline docs/kind.md:467]:
```
Using a local registry is an alternative to deploy ovn-kubernetes updates while also being useful to deploy other local images
```
I will admit though that I am missing the usefulness of hosting both ovn-k and other images in the local registry other than convenience. :thinking:

File: contrib/kind.sh [hunk 11]
```diff
@@ -593,29 +597,72 @@ set_cluster_cidr_ip_families() {
...
+  ovn_num_worker=${KIND_NUM_WORKER} \
+  cluster_log_level=${KIND_CLUSTER_LOGLEVEL:-4} \
+  j2 "${KIND_CONFIG}" -o "${KIND_CONFIG_LCL}"
 
   # Create KIND cluster. For additional debug, add '--verbosity <int>': 0 None .. 3 Debug
   if kind get clusters | grep ovn; then
     delete
   fi
+  
+  if [[ "${KIND_LOCAL_REGISTRY}" == true ]]; then
```
Comment (jcaamano) [inline contrib/kind.sh:651]: So hopefully re-using the existing flag `--local-kind-registry` with a slightly different meaning does not disrupt anyone's workflow but that's a thought I had.

**Review by jcaamano — CHANGES_REQUESTED** [review]

File: contrib/kind.sh [hunk 12]
```diff
@@ -593,29 +597,72 @@ set_cluster_cidr_ip_families() {
...
-    j2 "${KIND_CONFIG}" -o "${KIND_CONFIG_LCL}"
+  ovn_ip_family=${IP_FAMILY} \
+  ovn_ha=${OVN_HA} \
+  net_cidr=${NET_CIDR} \
+  svc_cidr=${SVC_CIDR} \
+  use_local_registy=${KIND_LOCAL_REGISTRY} \
+  dns_domain=${KIND_DNS_DOMAIN} \
+  ovn_num_master=${KIND_NUM_MASTER} \
+  ovn_num_worker=${KIND_NUM_WORKER} \
+  cluster_log_level=${KIND_CLUSTER_LOGLEVEL:-4} \
```
[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh:643]: the internal part of the registry can be "harcoded" without much issue the problematic part if the external port and it's already parameterized, this is what it's done at kind config template 

```
{%- if use_local_registy == "true"%}                                            
containerdConfigPatches:                                                        
  - |-                                                                          
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:5000"]     
      endpoint = ["http://kind-registry:5000"]                                  
{%- endif %}          
```

File: contrib/kind.sh [hunk 13]
```diff
@@ -783,8 +836,9 @@ install_ovn() {
 
   popd
 
-  # Force pod reload just the ones with golang containers
-  if [ "${KIND_CREATE}" == false ]; then
+  # When using internal registry force pod reload just the ones with 
+  # golang containers
+  if [ "${KIND_CREATE}" == false ] && [ "${KIND_LOCAL_REGISTRY}" == false ] ; then
```
[the remark this answers is not in the document]
Comment (qinqon) [reply contrib/kind.sh:841]: > Wouldn't it be better to always use the sha and avoid this reload?
I tried that with internal registry but it was not working, don't remember why.

> I suspect golang containers is a typo. Would it mean ovn-kubernetes containers?
At the kind deploy also OVS is deployed as container but restarting those breaks the cluster, I will rephrase this to non ovs containers.

File: docs/kind.md (hunk 10 above)
Comment (qinqon) [reply docs/kind.md:467]: The "kind load docker-image" is a kind ad-hoc solution that do not play well with projects that already use the typical "docker push", in fact kind team is planing on replacing it totally with local registry running at pod on the cluster [1]

[1] https://github.com/kubernetes-sigs/kind/issues/1213

File: contrib/kind.sh (hunk 11 above)
Comment (qinqon) [reply contrib/kind.sh:651]: If the local registry is already there is not started, I think that was how it was working before.

File: docs/kind.md (hunk 10 above)
Comment (jcaamano) [reply docs/kind.md:465]: nit: could you replace `Using the kind local registry...` with my suggestion and remove `At some scenarios use...`?

Comment (qinqon) [reply docs/kind.md]: Done
